runtime/onert/api/python/package/infer/session.py

```diff
@@ -89,20 +76,46 @@ def infer(
         """
         metrics: Dict[str, float] = {}
 
-        # Check if the session is prepared. If not, call prepare() and set_outputs() once.
-        if not self._prepared:
-            with self._time_block(metrics, 'prepare_time_ms', measure):
-                self.session.prepare()
-                self.set_outputs(self.session.output_size())
-                self._prepared = True
-
         # Verify that the number of provided inputs matches the session's expected input count.
         expected_input_size: int = self.session.input_size()
         if len(inputs_array) != expected_input_size:
             raise ValueError(
                 f"Expected {expected_input_size} input(s), but received {len(inputs_array)}."
             )
 
+        # Check if the session is prepared. If not, call prepare() and set_outputs() once.
+        if not self._prepared:
+            with self._time_block(metrics, 'prepare_time_ms', measure):
+                # On first call, fix any -1 dims to real input shapes and validate
+                original_infos = self.get_inputs_tensorinfo()
+                fixed_infos = []
+                for idx, info in enumerate(original_infos):
+                    input_shape = inputs_array[idx].shape
+                    new_dims = []
+                    # only the first `info.rank` entries matter
+                    for j, d in enumerate(info.dims[:info.rank]):
+                        if d == -1:
+                            # replace dynamic dim with actual incoming shape
+                            new_dims.append(input_shape[j])
+                        elif d == input_shape[j]:
+                            # static dim must match the provided array
+                            new_dims.append(d)
+                        else:
+                            raise ValueError(
+                                f"Input #{idx} dim {j} mismatch: "
+                                f"tensorinfo={d}, actual input shape={input_shape[j]}")
```

Comment:
Is it intended decision? (using error, not warning)

Reply:
I was waiting for your opinion for warning handling.